### DISCUSSION TITLE
how to draw the image in which the blocks are connected bidirectionaly?

### DISCUSSION THREAD

**Discussion by zhuzhzh** [plantuml/plantuml#1486] [Q&A]
Here is one example.

![image](https://github.com/plantuml/plantuml/assets/14127098/f1371607-beab-46b3-9b20-5f427d82d29b)

**Comment by Potherca** [comment]
Depending on what your needs are, you have three options:

| | | 
| --- | --- |
| `A <-> B ` | ![](https://www.plantuml.com/plantuml/png/SoWkIImgAStDuIfAJIv9p4lFILLmH68xSN4vAjZesW4P3gbvAK250000) |
| `A -> B` and `A <- B` | ![](https://www.plantuml.com/plantuml/png/SoWkIImgAStDuIfAJIv9p4lFILLmH68xSN4vAkZQAJW1AHjT8EMGcfS2L0u0) |
| `A -> B` and `B -> A` | ![](http://www.plantuml.com/plantuml/png/SoWkIImgAStDuIfAJIv9p4lFILLmH68xSN4vAkZQ0Mad80MKSaZDIm4g1m00) |

Or, to demonstrate all 3 options together:

<table><tr><td>

```plantuml
@startuml

rectangle A
rectangle B
rectangle C
rectangle D

A <-> B
B -> C
B <- C
C -> D
D -> C

@enduml
```
</td><td>

![](http://www.plantuml.com/plantuml/png/SoWkIImgAStDuU8gIaqkISnBpqbLS4HYEo6ndP7OBbnSZWeskdP01KuAGCePIDde0Ybd80yewm8Hv798pKi16Wy0)
</td></tr></table>

As pointed out by @The-Lum, a naive implementation will not get you the same look as your original image, because of how the arrow are positioned for containers. To get a more similar look, you would need to use hidden elements _inside_ the containers you are linking between:

![](http://www.plantuml.com/plantuml/png/VL91Q_iW7BtxAQGz1_v9U_WFnYZj5kXXhB3kD4QHn6HIg-MdOPRHxpwZsLIQLY_rlUTxxsUpK9g0xWuSgJqJHmBaWByHjgkc3UcuBgNGAyabuEdyVx-HOvdG51fIKQzsQqcgVGko4xM_e848PIsfq44ZeQPmdNTne-pP1mEjD14jlnKyVRhhzqXec28EVDKdwtVaX8augY9GtqeftRgT4jm8WFnAMSkuZ9KK3MjBnYb2ORuDI7CQyfriKZaTLgpdjHGJJ1IkLAHOSLdjyLeqqdCiSFQofaOFep6SmVidgsigFZ9iyW5M-3vx-Bl6xCaUXvj14M7dkF6oV8xxjbJxoG9qx28e2BndIfeYgPYQcHdE-0YPzdbysyPzRBtTMYXTN4Bv9VGlGDPdjttPk3SoyxXMK6DBXjEe9ayq-QYcm2ubEFcd6HeaL7wTAbBAFWfoVnpFB1n_RgOt0OkP0fdPx-yvMb1HTmV-0m00)

<details><summary>Expand to see diagram code</summary>

```plantuml
@startuml
skinparam {
    defaultFontColor #484848

    interface {
        BackgroundColor transparent
        BorderColor transparent
        FontColor transparent
    }

    rectangle {
        BackgroundColor #E9E9E9
        BorderColor #484848
        FontStyle plain
    }

    card {
        BackgroundColor #FFFFFF
        BorderColor #484848
    }

    arrowColor #484848
}

card ConfigFile

rectangle Proxy {
    card "config singleton" as cs
    card "Clock Info" as ci
    card context

    ci -r[hidden]- x1

    cs <-> ci
    ci <--> context
    cs <--> context
}

rectangle BFM {
    card getConfig
    card putConfig
    getConfig <--> putConfig
    x2 -r[hidden]- getConfig
    x3 -l[hidden]- getConfig
}

rectangle DUT {
    card BFM_Inst2
    card BFM_Inst1
    card BFM_Inst0
    card DUT_TOP

   x4 -r[hidden]- BFM_Inst0 
   BFM_Inst0 -r[hidden]- BFM_Inst1 
   BFM_Inst1 -r[hidden]- BFM_Inst2 

    BFM_Inst0 <--> DUT_TOP
    BFM_Inst1 <--> DUT_TOP
    BFM_Inst2 <--> DUT_TOP
}


cs <-l---> ConfigFile
x1 <-r-> x2
x3 <-r-> x4
@enduml
```

</details>

**Comment by The-Lum** [comment]
Hi @zhuzhzh, @Potherca, and all,

Here is an attempt:
```plantuml
'skinparam linetype ortho

card "Config File" as cf

rectangle Proxy as p {
  card p1 as "config singleton"
  card p2 as "Clock Info"
  card p3 as "context"
}

rectangle "BFM" as bfm {
  card "getConfig" as getC
  card "putConfig" as putC
}


rectangle "DUT" as dut {
  card "BFM_Inst0" as d0
  card "BFM_Inst1" as d1
  card "BFM_Inst2" as d2
  card "DUT_TOP" as d
}

cf <-> p1
p1 <-> p2
p1 <-> p3
p2 <--> p3

p <--> bfm
getC <--> putC

bfm <--> dut
d0 <--> d
d1 <--> d
d2 <--> d
```
>[![](https://www.planttext.com/api/plantuml/svg/PP0zRm8n38Pt_mehIoUaksD57Kg5n83A03DAyt6DE98e5oGGwdzld5p3IMol_TX-RR-DPsCz3_o2Wx4gthr25-AF0n0yI6HhPxNfSMC6nP2FA3H0K29osmyAzy7TxfJs-034tE9RIZ1H6aT3XT5PLdcN-NfmuenRgzsJBFyxexf51hzpA_QnsUKDllNbQSPw5SkAcL5KaR_E4KKqSJxoytZ8J5xZR6GoEctj69l2cXVG5j2-WAw0he9aS3fyxKkQt8N6rU8z_GZIcxBggbf2UirgKIJu8jErG5TDW8u0-a0EqzuWcqc3RAkQviW_)](https://www.planttext.com/?text=PP0zRm8n38Pt_mehIoUaksD57Kg5n83A03DAyt6DE98e5oGGwdzld5p3IMol_TX-RR-DPsCz3_o2Wx4gthr25-AF0n0yI6HhPxNfSMC6nP2FA3H0K29osmyAzy7TxfJs-034tE9RIZ1H6aT3XT5PLdcN-NfmuenRgzsJBFyxexf51hzpA_QnsUKDllNbQSPw5SkAcL5KaR_E4KKqSJxoytZ8J5xZR6GoEctj69l2cXVG5j2-WAw0he9aS3fyxKkQt8N6rU8z_GZIcxBggbf2UirgKIJu8jErG5TDW8u0-a0EqzuWcqc3RAkQviW_)

_But that's not so perfect for lines between horizontal rectangles..._ 🙅 👎 
Regards

**Reply by zhuzhzh** [reply]
maybe one attribute can be added to define the alignment/direction of elements. I don't know if it's feasible.

**Comment by zhuzhzh** [comment]
@Potherca @The-Lum  Thanks so much!
